### DISCUSSION TITLE
Testing with Jest

### DISCUSSION THREAD

**Discussion by Raphael0010** [trpc/trpc#1740] [Q&A]
Hi, there is a way to test trpc query with Jest ?.

I'm actually trying to do that : 

```ts
import { trpcClient } from '@project/trpc/dist/trpc'

test('getAllBids', () => {
  const bids = trpcClient.query('bid.getAll')
  expect(bids).toBe('test')
})
```

But I got this error : "Argument of type 'string' is not assignable to parameter of type 'never'."
But when I use it on Next App it's work.
I see nothing about testing in the documentation.

**Comment by LouisHaftmann** [comment]
Have you tried using a router caller?

```tsx
appRouter.createCaller(/* context object */).query('bid.getAll')
```

**Reply by uxduck** [reply]
It should look a bit better now instead of using strings. I've been writing tests like:

```
    const caller = appRouter.createCaller({
      services,
      db,
      ...
    });
   const cats = await caller.cat.getAll();
   ...
```

**Comment by sachinraja** [comment]
You could also use the client if you [pass the type of your router as a type arg](https://trpc.io/docs/vanilla#initialize-a-trpc-client) and your trpc server is running during your tests. Hard to help more than that without a reproduction.

**Comment by KATT** [comment]
There is both jest and playwright in the starter

https://github.com/trpc/examples-next-prisma-starter/tree/main/src/server/routers

**Comment by Raphael0010** [comment]
The fact is that I'm using Next.js what should I do to mock req and res ? 

This is my context : 

```ts
export const createContext = async ({ req, res }: trpcNext.CreateNextContextOptions) => {
  return {
    req,
    res,
    prisma,
  }
}
```

**Reply by LouisHaftmann** [reply]
if you are not using 'req' or 'res' in your queries/mutations:

```tsx
appRouter.createCaller({} /* just pass an empty object */)

appRouter.createCaller({ prisma } /* or just prisma */)
```

**Reply by Raphael0010** [reply]
We using it

**Reply by LouisHaftmann** [reply]
then you need a trpc client. the caller just calls the route resolver and doesn't actually make any http requests. could you please provide a bit more context (code, examples) regarding your original request?

**Reply by Raphael0010** [reply]
The trpc client still saying : "Argument of type 'string' is not assignable to parameter of type 'never'." this is what I do in first.

About context, it's just basic next.js project with trpc and jest ....

**Reply by KATT** [reply]
I usually have a distinct `createContextInner`-function for tests that doesn't need the req.

```ts
// context.ts

/**
 * Creates our context based on passed in arguments for cases where the `req` isn't available
 */
export async function createContextInner(opts: {
  session: Session | null;
}) {
  const { session } = opts;

  return {
    session,
  };
}

export type Context = inferAsyncReturnType<typeof createContextInner>;

/**
 * Creates context for an incoming request
 * @link https://trpc.io/docs/context
 */
export async function createContext(opts: trpcNext.CreateNextContextOptions) {
  const { req } = opts;

  const session = await getSession({ req });

  const ctx = await createContextInner({
    session,
  });
  return ctx;
}


```

**Reply by KATT** [reply]
The `createContextInner` can easily be called from integration tests and you for instance mock the current session more easily, or replace your default logger with a mock logger, etc

**Reply by KATT** [reply]
See https://github.com/trpc/trpc/pull/1746

**Reply by Raphael0010** [reply]
Still having issue with the typing ...

```ts
const context = await createContextInner()
const caller = appRouter.createCaller(context)
const bids = caller.query('bid.getAll')
```

Got this error : 
```
__tests__/trpc.test.ts:7:29 - error TS2345: Argument of type 'string' is not assignable to parameter of type 'never'.
```

This is my context : 
```ts
export const createContextInner = async () => {
  return {
    prisma,
  }
}
```

**Reply by KATT** [reply]
This might help

https://github.com/trpc/examples-next-prisma-starter/blob/main/jest.config.ts#L40

**Reply by Raphael0010** [reply]
Still nothing :(

**Reply by KATT** [reply]
Okay, try to compare https://github.com/trpc/examples-next-prisma-starter to your setup, it's impossible for us to debug

**Reply by Raphael0010** [reply]
Ok, I have tried with fresh turbo repo with next, and it's work. I know now why i don't have the typing in my tests it's because I have made a trpc packages built with ts-up, so why trpc need to be built in next app ?

**Comment by gaurangrshah** [comment]
Take a look at the setup used in the BisonApp project: https://github.com/echobind/bisonapp/blob/c6c7f3604c99d5f27bf351f9a8c6f3e980e2af89/packages/create-bison-app/template/tests/helpers/trpcRequest.ts

I've got this working using their approach: 
```tsx
  test('🟢', async () => {
    const response = await trpcRequest().example.hello({
      text: `from ${process.env.NEXT_PUBLIC_APP_ENV}`,
    });

    expect(response.greeting).toBe('Hello from test');
  });
```

**Comment by swport** [comment]
why isn't testing with jest part of the official docs. why do authors love to see new comers struggle with such things?? pathetic.

**Reply by jussisaurio** [reply]
This isn't an attitude that's going to get people to help you. In any case, describe an example you'd like to test, and maybe someone might still decide to help.

**Reply by crummy** [reply]
isn't this link, which got from an earlier post, an official example of testing with jest? https://github.com/trpc/examples-next-prisma-starter/blob/main/src/server/routers/post.test.ts
